### ISSUE TITLE
Unable to run nmap through proxychains-ng/proxychains4

### ISSUE THREAD

**Issue by cyberstx** [rofl0r/proxychains-ng#103]
Hello!

After installing & configuring proxychains-ng I'm unable to do anything with Nmap, I can still use proxychains-ng with tor service tho.

Getting the error :

"nmap: netutil.cc:1350: int collect_dnet_interfaces(const intf_entry*, void*): Assertion `rc == 0' failed.
Aborted"

Searched over the net without any luck.
I'm on Kali 2.0, fully updated.

Thanks.

**Comment by rofl0r** [comment]
are you aware of the fact that proxychains is TCP only ?
UDP,IDCMP,raw packets etc are not supported. simple syn-scans and other TCP based scans should work, though.

**Comment by cyberstx** [comment]
Yes I know it's only TCP, I'm trying with those command lines :

proxychains4 nmap -sS <IP>
proxychains4 nmap -sT <IP>

None of them are working.
It's working with proxychains tho.

**Comment by rofl0r** [comment]
which nmap version ?
does it work if you specify the network interface to use, for example
```
./proxychains4 -f /tmp/proxychains.conf nmap -Ss 192.168.1.10 -e eth0
```

**Comment by cyberstx** [comment]
Ok so, got some news.
I'm using nmap version 7.01.
Your command did not work - BUT - I found the problem. I got the error when I let the "proxy_dns" option on (uncommented). Without this option it works perfectly. I'm changing the DNS in my resolv.conf to openNIC ones so I think this way I'm still able to prevents DNS leak but not through proxychains right?

**Comment by iv4nuk0** [comment]
Hello, I have the same problem and I can not answer, and tried the above command with the interface and does not work and change the DNS does not work, use the nmap version 7.01 and proxychains 4, the error is this:
Starting Nmap 7.01 ( https://nmap.org ) at 2016-02-08 15:10 CET
nmap: netutil.cc:1350: int collect_dnet_interfaces(const intf_entry*, void*): Assertion `rc == 0' failed.
Abortado
thx

**Comment by rofl0r** [comment]
yeah, nmap has some stupid (superfluous) code in it that tries to determine which network interface to use to send out packets, even if the user uses tcp options, so it could just unconditionally connect() and leave the decision to the OS.
please file a bug against nmap.
in the meantime there exist 3 options to workaround it on the proxychains side:
1) do not use a dns name, but a raw ipv4 address
2) or disable proxy_dns in the config
3) or do not use proxychains at all but nmap's integrated options to use a SOCKS proxy.

**Comment by bobbuilder123** [comment]
How do 1 remove proxy_dns?

**Comment by Dreamwhite** [comment]
@cyberstx   I have the same problem.And I close the "proxy-dns",then nmap start to work.
Thank you!

**Comment by leebaird** [comment]
Still have the same issue.

**Comment by rofl0r** [comment]
to get this fixed, we need to make nmap developers aware of the problem and get them involved. then we may be able to find a mutual solution. so someone needs to stand up and get in touch with them (maybe with a link to this thread).

**Comment by radmanxu** [comment]
When you close the "proxy-dns" ,it looks like works ,but you will use your true ip to scan !!!

**Comment by radmanxu** [comment]
https://security.stackexchange.com/questions/122561/how-to-use-nmap-through-proxychains

**Comment by keshav3** [comment]
Same nmap: netutil.cc:3285: int route_dst_netlink(const sockaddr_storage*, route_nfo*, const char*, const sockaddr_storage*): Assertion `p != NULL' failed.
Aborted tell me plz

**Comment by rofl0r** [comment]
report it on nmap mailing list

**Comment by keshav3** [comment]
> report it on nmap mailing list

Ok bro

**Comment by Green-m** [comment]
In my case, it worked when I disable the ping host discovery with `-Pn` like this:

```
proxychains4 -q  nmap -sT 192.168.1.10 -Pn
```

**Comment by Samshingenge** [comment]
I had the same problem " uncommented proxy_dns then it works in the proxychains.conf file.

**Comment by rofl0r** [comment]
> uncommented proxy_dns then it works

i guess you meant to say you commented it out. did anyone ever report this issue to nmap developers ?

**Comment by 4390c336** [comment]
the main issues is that you are running nmap as root .
and when nmap is running as root it will try to use low level raw sockets.
fix your problem you have 2 options : 
* run nmap is an unprivileged user
* add `--unprivileged ` to your command , or set `NMAP_UNPRIVILEGED` env variable to save you from typing it every time
